// @kitware/vtk.js@27.3.1/Filters/General.js
import vtkAppendPolyData from './General/AppendPolyData.js';
import vtkCalculator from './General/Calculator.js';
import vtkClipClosedSurface from './General/ClipClosedSurface.js';
import vtkClosedPolyLineToSurfaceFilter from './General/ClosedPolyLineToSurfaceFilter.js';
import vtkContourTriangulator from './General/ContourTriangulator.js';
import vtkImageCropFilter from './General/ImageCropFilter.js';
import vtkImageMarchingCubes from './General/ImageMarchingCubes.js';
import vtkImageMarchingSquares from './General/ImageMarchingSquares.js';
import vtkImageOutlineFilter from './General/ImageOutlineFilter.js';
import vtkImageSliceFilter from './General/ImageSliceFilter.js';
import vtkImageStreamline from './General/ImageStreamline.js';
import vtkLineFilter from './General/LineFilter.js';
import vtkMoleculeToRepresentation from './General/MoleculeToRepresentation.js';
import vtkOBBTree from './General/OBBTree.js';
import vtkOutlineFilter from './General/OutlineFilter.js';
import vtkPaintFilter from './General/PaintFilter.js';
import vtkScalarToRGBA from './General/ScalarToRGBA.js';
import vtkTriangleFilter from './General/TriangleFilter.js';
import vtkTubeFilter from './General/TubeFilter.js';
import vtkWarpScalar from './General/WarpScalar.js';
import vtkWindowedSincPolyDataFilter from './General/WindowedSincPolyDataFilter.js';

var General = {
  vtkAppendPolyData: vtkAppendPolyData,
  vtkCalculator: vtkCalculator,
  vtkClipClosedSurface: vtkClipClosedSurface,
  vtkClosedPolyLineToSurfaceFilter: vtkClosedPolyLineToSurfaceFilter,
  vtkContourTriangulator: vtkContourTriangulator,
  vtkImageCropFilter: vtkImageCropFilter,
  vtkImageMarchingCubes: vtkImageMarchingCubes,
  vtkImageMarchingSquares: vtkImageMarchingSquares,
  vtkImageOutlineFilter: vtkImageOutlineFilter,
  vtkImageSliceFilter: vtkImageSliceFilter,
  vtkImageStreamline: vtkImageStreamline,
  vtkLineFilter: vtkLineFilter,
  vtkMoleculeToRepresentation: vtkMoleculeToRepresentation,
  vtkOBBTree: vtkOBBTree,
  vtkOutlineFilter: vtkOutlineFilter,
  vtkPaintFilter: vtkPaintFilter,
  vtkScalarToRGBA: vtkScalarToRGBA,
  vtkTriangleFilter: vtkTriangleFilter,
  vtkTubeFilter: vtkTubeFilter,
  vtkWarpScalar: vtkWarpScalar,
  vtkWindowedSincPolyDataFilter: vtkWindowedSincPolyDataFilter
};

export { General as default };
