// @kitware/vtk.js@27.3.1/Filters/General/OBBTree/helper.js
import { CellType } from '../../../Common/DataModel/CellTypes/Constants.js';
import vtkOBJReader from '../../../IO/Misc/OBJReader.js';

/**
 * Get the correct point ID from a cell id
 * @param {Array} cellPtsIds
 * @param {CellType} type
 * @param {Number} idx
 * @returns {Object} Contains three point's id of cells as 'ptId0', 'ptId1', 'ptId2'
 */

function getCellTriangles(cellPtsIds, type, idx) {
  var ptId0 = -1;
  var ptId1 = -1;
  var ptId2 = -1;
  var cellListLength = cellPtsIds.length;

  switch (type) {
    case CellType.VTK_TRIANGLE:
    case CellType.VTK_POLYGON:
    case CellType.VTK_QUAD:
      {
        if (idx > cellListLength) break;
        ptId0 = cellPtsIds[0];
        ptId1 = cellPtsIds[idx + 1];
        ptId2 = cellPtsIds[idx + 2];
        break;
      }

    case CellType.VTK_TRIANGLE_STRIP:
      {
        // eslint-disable-next-line no-bitwise
        var idx1 = idx + 1 + (idx & 1); // eslint-disable-next-line no-bitwise

        var idx2 = idx + 2 - (idx & 1);
        if (idx1 > cellListLength || idx2 > cellListLength) break;
        ptId0 = cellPtsIds[idx];
        ptId1 = cellPtsIds[idx1];
        ptId2 = cellPtsIds[idx2];
        break;
      }

    default:
      ptId0 = -1;
      ptId1 = -1;
      ptId2 = -1;
      break;
  }

  return {
    ptId0: ptId0,
    ptId1: ptId1,
    ptId2: ptId2
  };
}
/**
 * Concatenate second typed array to the first typed array.
 * @param {TypedArray} first
 * @param {TypedArray} second Must be of the same type as first
 * @return {TypedArray}
 */

function pushArray(first, second) {
  var firstLength = first.length;
  var result = new first.constructor(firstLength + second.length);
  result.set(first);
  result.set(second, firstLength);
  return result;
}
/**
 * Load an obj with point's colors
 *
 * @param {string} url path to the OBJ file
 * @return Promise
 * ---> success : Return vtkPolyData
 * ---> failed : Error message
 */

function loadOBJ(url) {
  return new Promise(function (resolve, reject) {
    var reader = vtkOBJReader.newInstance();
    reader.setUrl(url).then(function () {
      var data = reader.getOutputData();
      resolve(data);
    }, function () {
      // eslint-disable-next-line prefer-promise-reject-errors
      reject('Error when loading ', url);
    });
  });
}

export { getCellTriangles, loadOBJ, pushArray };
