// @kitware/vtk.js@27.3.1/Widgets/Representations/ArrowHandleRepresentation.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import macro from '../../macros.js';
import vtkActor from '../../Rendering/Core/Actor.js';
import vtkArrow2DSource from '../../Filters/Sources/Arrow2DSource.js';
import vtkGlyphRepresentation from './GlyphRepresentation.js';
import vtkMatrixBuilder from '../../Common/Core/MatrixBuilder.js';
import vtkPixelSpaceCallbackMapper from '../../Rendering/Core/PixelSpaceCallbackMapper.js';
import vtkConeSource from '../../Filters/Sources/ConeSource.js';
import vtkSphereSource from '../../Filters/Sources/SphereSource.js';
import vtkCircleSource from '../../Filters/Sources/CircleSource.js';
import vtkCubeSource from '../../Filters/Sources/CubeSource.js';
import vtkViewFinderSource from '../../Filters/Sources/ViewFinderSource.js';
import Constants from '../Widgets3D/LineWidget/Constants.js';
import { mat4, mat3, vec3 } from 'gl-matrix';
import { RenderingTypes } from '../Core/WidgetManager/Constants.js';
import { OrientationModes } from '../../Rendering/Core/Glyph3DMapper/Constants.js';
import { allocateArray } from './WidgetRepresentation.js';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var ShapeType = Constants.ShapeType,
    Shapes2D = Constants.Shapes2D,
    ShapesOrientable = Constants.ShapesOrientable; // ----------------------------------------------------------------------------
// vtkArrowHandleRepresentation methods
// ----------------------------------------------------------------------------

function vtkArrowHandleRepresentation(publicAPI, model) {
  // Set our className
  model.classHierarchy.push('vtkArrowHandleRepresentation');

  var superClass = _objectSpread({}, publicAPI); // --------------------------------------------------------------------------
  // Internal polydata dataset
  // --------------------------------------------------------------------------

  /**
   * Set the shape for the glyph according to lineWidget state inputs
   */


  function createGlyph(shape) {
    var _representationToSour;

    var representationToSource = (_representationToSour = {}, _defineProperty(_representationToSour, ShapeType.STAR, {
      builder: vtkArrow2DSource,
      initialValues: {
        shape: 'star',
        height: 0.6
      }
    }), _defineProperty(_representationToSour, ShapeType.ARROWHEAD3, {
      builder: vtkArrow2DSource,
      initialValues: {
        shape: 'triangle'
      }
    }), _defineProperty(_representationToSour, ShapeType.ARROWHEAD4, {
      builder: vtkArrow2DSource,
      initialValues: {
        shape: 'arrow4points'
      }
    }), _defineProperty(_representationToSour, ShapeType.ARROWHEAD6, {
      builder: vtkArrow2DSource,
      initialValues: {
        shape: 'arrow6points'
      }
    }), _defineProperty(_representationToSour, ShapeType.CONE, {
      builder: vtkConeSource,
      initialValues: {
        direction: [0, 1, 0]
      }
    }), _defineProperty(_representationToSour, ShapeType.SPHERE, {
      builder: vtkSphereSource
    }), _defineProperty(_representationToSour, ShapeType.CUBE, {
      builder: vtkCubeSource,
      initialValues: {
        xLength: 0.8,
        yLength: 0.8,
        zLength: 0.8
      }
    }), _defineProperty(_representationToSour, ShapeType.DISK, {
      builder: vtkCircleSource,
      initialValues: {
        resolution: 30,
        radius: 0.5,
        direction: [0, 0, 1],
        lines: false,
        face: true
      }
    }), _defineProperty(_representationToSour, ShapeType.CIRCLE, {
      builder: vtkCircleSource,
      initialValues: {
        resolution: 30,
        radius: 0.5,
        direction: [0, 0, 1],
        lines: true,
        face: false
      }
    }), _defineProperty(_representationToSour, ShapeType.VIEWFINDER, {
      builder: vtkViewFinderSource,
      initialValues: {
        radius: 0.1,
        spacing: 0.3,
        width: 1.4
      }
    }), _defineProperty(_representationToSour, ShapeType.NONE, {
      builder: vtkSphereSource
    }), _representationToSour);
    var rep = representationToSource[shape];
    return rep.builder.newInstance(rep.initialValues);
  } // --------------------------------------------------------------------------
  // Generic rendering pipeline
  // --------------------------------------------------------------------------
  // displayActors and displayMappers are used to render objects in HTML,
  // allowing objects to be 'rendered' internally in a VTK scene without
  // being visible on the final output.


  model.displayMapper = vtkPixelSpaceCallbackMapper.newInstance();
  model.displayActor = vtkActor.newInstance({
    parentProp: publicAPI
  }); // model.displayActor.getProperty().setOpacity(0); // don't show in 3D

  model.displayActor.setMapper(model.displayMapper);
  model.displayMapper.setInputConnection(publicAPI.getOutputPort());
  publicAPI.addActor(model.displayActor);
  model.alwaysVisibleActors = [model.displayActor]; // --------------------------------------------------------------------------

  publicAPI.setGlyphResolution = macro.chain(publicAPI.setGlyphResolution, function (r) {
    return model._pipeline.glyph.setPhiResolution(r);
  }, function (r) {
    return model._pipeline.glyph.setThetaResolution(r);
  }); // --------------------------------------------------------------------------

  function callbackProxy(coords) {
    if (model.displayCallback) {
      var filteredList = [];
      var states = publicAPI.getRepresentationStates();

      for (var i = 0; i < states.length; i++) {
        if (states[i].getActive()) {
          filteredList.push(coords[i]);
        }
      }

      if (filteredList.length) {
        model.displayCallback(filteredList);
        return;
      }
    }

    model.displayCallback();
  }

  publicAPI.setDisplayCallback = function (callback) {
    model.displayCallback = callback;
    model.displayMapper.setCallback(callback ? callbackProxy : null);
  }; // --------------------------------------------------------------------------


  publicAPI.is2DShape = function () {
    return Shapes2D.includes(model.shape);
  };

  publicAPI.isOrientableShape = function () {
    return ShapesOrientable.includes(model.shape);
  };
  /**
   * Returns the orientation matrix to align glyph on model.orientation.
   * */


  function getOrientationRotation(viewMatrixInv) {
    var displayOrientation = new Float64Array(3);
    var baseDir = [0, 1, 0];
    vec3.transformMat3(displayOrientation, model.orientation, viewMatrixInv);
    displayOrientation[2] = 0;
    var displayMatrix = vtkMatrixBuilder.buildFromDegree().rotateFromDirections(baseDir, displayOrientation).getMatrix();
    var displayRotation = new Float64Array(9);
    mat3.fromMat4(displayRotation, displayMatrix);
    return displayRotation;
  }

  function getCameraFacingRotation(scale3, displayRotation, viewMatrix) {
    var rotation = new Float64Array(9);
    mat3.multiply(rotation, viewMatrix, displayRotation);
    vec3.transformMat3(scale3, scale3, rotation);
    return rotation;
  }
  /**
   * Computes the rotation matrix of the glyph. There are 2 rotations:
   *  - a first rotation to be oriented along model.rotation
   *  - an optional second rotation to face the camera
   * @param {vec3} scale3 Scale of the glyph, rotated when glyph is rotated.
   */


  function getGlyphRotation(scale3) {
    var shouldFaceCamera = model.faceCamera === true || model.faceCamera == null && publicAPI.is2DShape();
    var viewMatrix = new Float64Array(9);
    mat3.fromMat4(viewMatrix, model.viewMatrix);
    var viewMatrixInv = mat3.identity(new Float64Array(9));

    if (shouldFaceCamera) {
      mat3.invert(viewMatrixInv, viewMatrix);
    }

    var orientationRotation = null;

    if (publicAPI.isOrientableShape()) {
      orientationRotation = getOrientationRotation(viewMatrixInv);
    } else {
      orientationRotation = mat3.identity(new Float64Array(9));
    }

    if (shouldFaceCamera) {
      orientationRotation = getCameraFacingRotation(scale3, orientationRotation, viewMatrix);
    }

    return orientationRotation;
  }

  function applyOrientation(polyData, states) {
    model._pipeline.mapper.setOrientationArray('orientation');

    model._pipeline.mapper.setOrientationMode(OrientationModes.MATRIX);

    var orientation = allocateArray(polyData, 'orientation', states.length, 'Float32Array', 9).getData();
    var defaultScale3 = [1, 1, 1];

    for (var i = 0; i < states.length; ++i) {
      var _states$i$getScale, _states$i$getScale2, _states$i;

      var scale3 = (_states$i$getScale = (_states$i$getScale2 = (_states$i = states[i]).getScale3) === null || _states$i$getScale2 === void 0 ? void 0 : _states$i$getScale2.call(_states$i)) !== null && _states$i$getScale !== void 0 ? _states$i$getScale : defaultScale3;
      var rotation = getGlyphRotation(scale3);
      orientation.set(rotation, 9 * i);
    }
  }

  publicAPI.setDirection(applyOrientation);
  publicAPI.setNoOrientation(applyOrientation);

  publicAPI.requestData = function (inData, outData) {
    var _publicAPI$getReprese;

    // FIXME: shape should NOT be mixin, but a representation property.
    var shape = (_publicAPI$getReprese = publicAPI.getRepresentationStates(inData[0])[0]) === null || _publicAPI$getReprese === void 0 ? void 0 : _publicAPI$getReprese.getShape();
    var shouldCreateGlyph = model._pipeline.glyph == null;

    if (model.shape !== shape && Object.values(ShapeType).includes(shape)) {
      model.shape = shape;
      shouldCreateGlyph = true;
    }

    if (shouldCreateGlyph && model.shape) {
      model._pipeline.glyph = createGlyph(model.shape);

      model._pipeline.mapper.setInputConnection(model._pipeline.glyph.getOutputPort(), 1);
    }

    return superClass.requestData(inData, outData);
  };

  publicAPI.updateActorVisibility = function () {
    var renderingType = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : RenderingTypes.FRONT_BUFFER;
    var ctxVisible = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;
    var handleVisible = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;
    var hasValidState = publicAPI.getRepresentationStates().length > 0;
    superClass.updateActorVisibility(renderingType, ctxVisible, handleVisible && hasValidState);
  };
} // ----------------------------------------------------------------------------
// Object factory
// ----------------------------------------------------------------------------

/**
 *  'shape' default value is used first time 'shape' mixin is invalid.
 *  'faceCamera' controls wether the glyph should face camera or not:
 *    - null or undefined to leave it to shape type (i.e. 2D are facing camera,
 *    3D are not)
 *    - true to face camera
 *    - false to not face camera
 */


function defaultValues(initialValues) {
  return _objectSpread({
    faceCamera: null,
    orientation: [1, 0, 0],
    shape: ShapeType.SPHERE,
    viewMatrix: mat4.identity(new Float64Array(16))
  }, initialValues);
} // ----------------------------------------------------------------------------


function extend(publicAPI, model) {
  var initialValues = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
  Object.assign(model, defaultValues(initialValues));
  vtkGlyphRepresentation.extend(publicAPI, model, initialValues);
  macro.setGetArray(publicAPI, model, ['visibilityFlagArray'], 2);
  macro.setGetArray(publicAPI, model, ['orientation'], 3);
  macro.setGetArray(publicAPI, model, ['viewMatrix'], 16);
  macro.setGet(publicAPI, model, ['faceCamera']); // Object specific methods

  vtkArrowHandleRepresentation(publicAPI, model);
} // ----------------------------------------------------------------------------

var newInstance = macro.newInstance(extend, 'vtkArrowHandleRepresentation'); // ----------------------------------------------------------------------------

var vtkArrowHandleRepresentation$1 = {
  newInstance: newInstance,
  extend: extend
};

export { vtkArrowHandleRepresentation$1 as default, extend, newInstance };
