// @kitware/vtk.js@27.3.1/Rendering/Misc/SynchronizableRenderWindow/BehaviorManager/CameraSynchronizer.js
import { newInstance as newInstance$1, obj, setGet, setGetArray, chain } from '../../../../macros.js';
import { l as normalize } from '../../../../Common/Core/Math/index.js';

var MODE_RESET_CAMERA = 'resetcamera';
var MODE_ORIENTATION = 'orientation';
var MODE_SAME = 'same';
var SynchronizationMode = {
  MODE_RESET_CAMERA: MODE_RESET_CAMERA,
  MODE_ORIENTATION: MODE_ORIENTATION,
  MODE_SAME: MODE_SAME
};

function vtkCameraSynchronizer(publicAPI, model) {
  model.classHierarchy.push('vtkCameraSynchronizer');
  var cameraState = new Float64Array(9);
  var direction = new Float64Array(3);
  var subscriptions = [];

  function updateListeners() {
    while (subscriptions.length) {
      subscriptions.pop().unsubscribe();
    }

    if (!model.srcRenderer || !model.dstRenderer) {
      return;
    }

    var srcCamera = model.srcRenderer.getActiveCamera();
    var interactor = model.srcRenderer.getRenderWindow().getInteractor();
    subscriptions.push(srcCamera.onModified(function () {
      if (!interactor.isAnimating()) {
        publicAPI.update();
      }
    }));
    subscriptions.push(interactor.onAnimation(publicAPI.update));
    subscriptions.push(interactor.onEndAnimation(publicAPI.update));
  } // Update listeners automatically


  model._srcRendererChanged = updateListeners;
  model._dstRendererChanged = updateListeners;

  function updatePreviousValues(position, focalPoint, viewUp) {
    if (cameraState[0] !== position[0] || cameraState[1] !== position[1] || cameraState[2] !== position[2] || cameraState[3] !== focalPoint[0] || cameraState[4] !== focalPoint[1] || cameraState[5] !== focalPoint[2] || cameraState[6] !== viewUp[0] || cameraState[7] !== viewUp[1] || cameraState[8] !== viewUp[2]) {
      cameraState[0] = position[0];
      cameraState[1] = position[1];
      cameraState[2] = position[2];
      cameraState[3] = focalPoint[0];
      cameraState[4] = focalPoint[1];
      cameraState[5] = focalPoint[2];
      cameraState[6] = viewUp[0];
      cameraState[7] = viewUp[1];
      cameraState[8] = viewUp[2];
      return cameraState;
    }

    return false;
  }

  publicAPI.update = function () {
    if (!model.active || !model.srcRenderer || !model.dstRenderer) {
      return;
    }

    var srcCamera = model.srcRenderer.getActiveCamera();
    var dstCamera = model.dstRenderer.getActiveCamera();
    var position = srcCamera.getReferenceByName('position');
    var focalPoint = srcCamera.getReferenceByName('focalPoint');
    var viewUp = srcCamera.getReferenceByName('viewUp');
    var change = updatePreviousValues(position, focalPoint, viewUp);

    if (!change) {
      return;
    }

    if (model.mode === MODE_ORIENTATION) {
      direction[0] = change[0] - change[3];
      direction[1] = change[1] - change[4];
      direction[2] = change[2] - change[5];
      normalize(direction);
      dstCamera.setPosition(model.focalPoint[0] + model.distance * direction[0], model.focalPoint[1] + model.distance * direction[1], model.focalPoint[2] + model.distance * direction[2]);
      dstCamera.setFocalPoint(model.focalPoint[0], model.focalPoint[1], model.focalPoint[2]);
      dstCamera.setViewUp(change[6], change[7], change[8]);
    } else {
      dstCamera.setPosition(change[0], change[1], change[2]);
      dstCamera.setFocalPoint(change[3], change[4], change[5]);
      dstCamera.setViewUp(change[6], change[7], change[8]);
    }

    if (model.mode === MODE_RESET_CAMERA) {
      model.dstRenderer.resetCamera();
    }
  };

  publicAPI.delete = chain(function () {
    return publicAPI.setSrcRenderer(null);
  }, // Will clear any listener
  publicAPI.delete); // If src/dstRenderer provided at constructor register listeners

  updateListeners();
}

var DEFAULT_VALUES = {
  mode: MODE_ORIENTATION,
  focalPoint: [0, 0, 0],
  distance: 6.8,
  active: true // srcRenderer: null,
  // dstRenderer: null,

};
function extend(publicAPI, model) {
  var initialValues = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
  Object.assign(model, DEFAULT_VALUES, initialValues); // Build VTK API

  obj(publicAPI, model);
  setGet(publicAPI, model, ['mode', 'active', 'srcRenderer', 'dstRenderer', 'distance']);
  setGetArray(publicAPI, model, ['focalPoint'], 3, 0); // Object methods

  vtkCameraSynchronizer(publicAPI, model);
} // ----------------------------------------------------------------------------

var newInstance = newInstance$1(extend, 'vtkCameraSynchronizer');
var vtkCameraSynchronizer$1 = {
  newInstance: newInstance,
  extend: extend,
  SynchronizationMode: SynchronizationMode
};

export { DEFAULT_VALUES, MODE_ORIENTATION, MODE_RESET_CAMERA, MODE_SAME, SynchronizationMode, vtkCameraSynchronizer$1 as default, extend, newInstance };
