// @kitware/vtk.js@27.3.1/Widgets/Widgets3D/SphereWidget.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { e as distance2BetweenPoints } from '../../Common/Core/Math/index.js';
import vtkAbstractWidgetFactory from '../Core/AbstractWidgetFactory.js';
import vtkPlanePointManipulator from '../Manipulators/PlaneManipulator.js';
import vtkSphereHandleRepresentation from '../Representations/SphereHandleRepresentation.js';
import vtkSphereContextRepresentation from '../Representations/SphereContextRepresentation.js';
import macro from '../../macros.js';
import widgetBehavior from './SphereWidget/behavior.js';
import stateGenerator from './SphereWidget/state.js';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function vtkSphereWidget(publicAPI, model) {
  model.classHierarchy.push('vtkSphereWidget');

  var superClass = _objectSpread({}, publicAPI);

  model.methodsToLink = ['scaleInPixels'];

  publicAPI.getRepresentationsForViewType = function (viewType) {
    return [{
      builder: vtkSphereHandleRepresentation,
      labels: ['moveHandle']
    }, {
      builder: vtkSphereHandleRepresentation,
      labels: ['centerHandle']
    }, {
      builder: vtkSphereHandleRepresentation,
      labels: ['borderHandle']
    }, {
      builder: vtkSphereContextRepresentation,
      labels: ['sphereHandle']
    }];
  }; // --- Public methods -------------------------------------------------------


  publicAPI.getRadius = function () {
    var h1 = model.widgetState.getCenterHandle();
    var h2 = model.widgetState.getBorderHandle();
    return Math.sqrt(distance2BetweenPoints(h1.getOrigin(), h2.getOrigin()));
  };

  publicAPI.setManipulator = function (manipulator) {
    superClass.setManipulator(manipulator);
    model.widgetState.getMoveHandle().setManipulator(manipulator);
    model.widgetState.getCenterHandle().setManipulator(manipulator);
    model.widgetState.getBorderHandle().setManipulator(manipulator);
  }; // --------------------------------------------------------------------------
  // initialization
  // --------------------------------------------------------------------------


  publicAPI.setManipulator(model.manipulator || vtkPlanePointManipulator.newInstance({
    useCameraNormal: true
  }));
}

var defaultValues = function defaultValues(initialValues) {
  return _objectSpread({
    behavior: widgetBehavior,
    widgetState: stateGenerator()
  }, initialValues);
};

function extend(publicAPI, model) {
  var initialValues = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
  Object.assign(model, defaultValues(initialValues));
  vtkAbstractWidgetFactory.extend(publicAPI, model, initialValues);
  macro.setGet(publicAPI, model, ['manipulator', 'widgetState']);
  vtkSphereWidget(publicAPI, model);
}
var newInstance = macro.newInstance(extend, 'vtkSphereWidget');
var index = {
  newInstance: newInstance,
  extend: extend
};

export { index as default, extend, newInstance };
