// @kitware/vtk.js@27.3.1/Widgets/Representations.js
import vtkCircleContextRepresentation from './Representations/CircleContextRepresentation.js';
import vtkContextRepresentation from './Representations/ContextRepresentation.js';
import vtkConvexFaceContextRepresentation from './Representations/ConvexFaceContextRepresentation.js';
import vtkCubeHandleRepresentation from './Representations/CubeHandleRepresentation.js';
import vtkGlyphRepresentation from './Representations/GlyphRepresentation.js';
import vtkHandleRepresentation from './Representations/HandleRepresentation.js';
import vtkImplicitPlaneRepresentation from './Representations/ImplicitPlaneRepresentation.js';
import vtkLineHandleRepresentation from './Representations/LineHandleRepresentation.js';
import vtkOutlineContextRepresentation from './Representations/OutlineContextRepresentation.js';
import vtkPolyLineRepresentation from './Representations/PolyLineRepresentation.js';
import vtkSphereHandleRepresentation from './Representations/SphereHandleRepresentation.js';
import vtkSplineContextRepresentation from './Representations/SplineContextRepresentation.js';
import vtkWidgetRepresentation from './Representations/WidgetRepresentation.js';

var Representations = {
  vtkCircleContextRepresentation: vtkCircleContextRepresentation,
  vtkContextRepresentation: vtkContextRepresentation,
  vtkConvexFaceContextRepresentation: vtkConvexFaceContextRepresentation,
  vtkCubeHandleRepresentation: vtkCubeHandleRepresentation,
  vtkGlyphRepresentation: vtkGlyphRepresentation,
  vtkHandleRepresentation: vtkHandleRepresentation,
  vtkImplicitPlaneRepresentation: vtkImplicitPlaneRepresentation,
  vtkLineHandleRepresentation: vtkLineHandleRepresentation,
  vtkOutlineContextRepresentation: vtkOutlineContextRepresentation,
  vtkPolyLineRepresentation: vtkPolyLineRepresentation,
  vtkSphereHandleRepresentation: vtkSphereHandleRepresentation,
  vtkSplineContextRepresentation: vtkSplineContextRepresentation,
  vtkWidgetRepresentation: vtkWidgetRepresentation
};

export { Representations as default };
