// @kitware/vtk.js@27.3.1/Widgets/Widgets3D.js
import vtkAngleWidget from './Widgets3D/AngleWidget.js';
import vtkDistanceWidget from './Widgets3D/DistanceWidget.js';
import vtkEllipseWidget from './Widgets3D/EllipseWidget.js';
import vtkImageCroppingWidget from './Widgets3D/ImageCroppingWidget.js';
import vtkImplicitPlaneWidget from './Widgets3D/ImplicitPlaneWidget.js';
import vtkInteractiveOrientationWidget from './Widgets3D/InteractiveOrientationWidget.js';
import vtkLineWidget from './Widgets3D/LineWidget.js';
import vtkPaintWidget from './Widgets3D/PaintWidget.js';
import vtkPolyLineWidget from './Widgets3D/PolyLineWidget.js';
import vtkRectangleWidget from './Widgets3D/RectangleWidget.js';
import vtkResliceCursorWidget from './Widgets3D/ResliceCursorWidget.js';
import vtkShapeWidget from './Widgets3D/ShapeWidget.js';
import vtkSplineWidget from './Widgets3D/SplineWidget.js';

var Widgets3D = {
  vtkAngleWidget: vtkAngleWidget,
  vtkDistanceWidget: vtkDistanceWidget,
  vtkEllipseWidget: vtkEllipseWidget,
  vtkImageCroppingWidget: vtkImageCroppingWidget,
  vtkImplicitPlaneWidget: vtkImplicitPlaneWidget,
  vtkInteractiveOrientationWidget: vtkInteractiveOrientationWidget,
  vtkLineWidget: vtkLineWidget,
  vtkPaintWidget: vtkPaintWidget,
  vtkPolyLineWidget: vtkPolyLineWidget,
  vtkRectangleWidget: vtkRectangleWidget,
  vtkResliceCursorWidget: vtkResliceCursorWidget,
  vtkShapeWidget: vtkShapeWidget,
  vtkSplineWidget: vtkSplineWidget
};

export { Widgets3D as default };
